Clear out account state and attempt recovery during authentication errrors

grigoryk opened the pull request:
Logic is now in place, but this still needs some tests and changelog entries.
Changes should be backwards-compatible.

Depends on https://github.com/mozilla/application-services/pull/1244

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

grigoryk commented:
Local testing notes:

```
I've added some hacky check to emit auth problems encountered during syncing, and changed accesstoken ttl to 1 in the fxalib, which means sync will encounter an auth problem right away.
testing this: sync does hit a problem right away, lets the state machine know, which successfully recovers by getting a new token
a legit password-changed case doesn't recover, and surfaces the problem to the user
```

grigoryk commented:
The only problem i'm observing right now is that `accountNeedsReauth` method can return false-positives while we're going through the recovery process. We can either add some flags to make it more robust, or ignore it for now.

jonalmeida commented on components/concept/sync/src/main/java/mozilla/components/concept/sync/OAuthAccount.kt:
```diff
@@ -5,6 +5,7 @@
 package mozilla.components.concept.sync
 
 import kotlinx.coroutines.Deferred
+import mozilla.components.support.base.observer.ObserverRegistry
```
lint will probably fail from this.

grigoryk commented on components/feature/sync/src/main/java/mozilla/components/feature/sync/StorageSync.kt:
```diff
@@ -58,6 +61,10 @@ class StorageSync(
         }
 
+        logger.debug("waiting 5s...")
+        delay(5000)
```
I’ll remove that. Left over from testing, needed to trigger a sync failure with a short lived access token.

grigoryk commented on components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt:
```diff
@@ -94,14 +97,14 @@ data class DeviceTuple(val name: String, val type: DeviceType, val capabilities:
     private val context: Context,
     private val config: Config,
-    private val scopes: Array<String>,
+    private val applicationScopes: Array<String>,
```
Yes, good call.

grigoryk replied:
Actually, leaving it as an `array` so that the public api does not change.

grigoryk commented on components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FirefoxAccount.kt:
```diff
@@ -177,17 +178,30 @@ class FirefoxAccount internal constructor(
      * Tries to fetch an access token for the given scope.
      *
      * @param singleScope Single OAuth scope (no spaces) for which the client wants access
+     * @param notifyAuthErrorRegistry Whether or not to notify [authErrorRegistry] of failures
```
It's actually quite necessary for the consumers of this API to be aware of the `authErrorRegistry`! They have no other way to learn about failures.

grigoryk commented:
Added tests, cleaned this up. Now we're just waiting for mozilla/application-services#1244 to land. Once that lands and we know what the shipping API shape is, I'll either tweak this PR, or add comments with links to issues to improve the API. I left a comment describing what I'd like that API to look like here: https://github.com/mozilla/application-services/pull/1244#discussion_r291740678

rfk commented on components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt:
```diff
@@ -483,7 +515,49 @@ open class FxaAccountManager(
+
+                        // fxalib maintains some internal token caches that need to be cleared whenever we
+                        // hit an auth problem. Call below makes that clean-up happen.
+                        account.onAccessTokenError()
```
I've filed https://github.com/mozilla/application-services/issues/1263 specifically for helping clean up the abstractions here.

grigoryk replied:
Linked in the comments.
